std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,247 @@ unittest
+    {
+        q = 1 - q;
+    }
+
+    double x = r.length * q;
+    real j;
+    real g = modf(x, j);
+    uint j_int = cast(uint) j;
+
+
+    static if (isRandomAccessRange!R)
+    {
+        if (g.approxEqual(0))
+        {
+            return (r[j_int - 1] + r[j_int]) / 2.0;
+        }
+        else
+        {
+            return r[j_int];
+        }
+    }
+    else
+    {
+        if (g.approxEqual(0))
+        {
+            ElementType!R prev_item;
+
+            r = r.drop(j_int - 1);
+            prev_item = r.front;
+            r.popFront;
+            return (r.front + prev_item) / 2.0;
+        }
+        else
+        {
+            return r.drop(j_int).front;
+        }
+    }
+
+    assert(0);
+}
+
+///
+@safe unittest
+{
+    import std.algorithm.sorting : sort;
+    import std.range : assumeSorted;
+
+    assert([1, 2, 3, 4].assumeSorted.quantile(0.25) == 1.5);
+    assert([1, 2, 3, 4].assumeSorted.quantile(0.5) == 2.5);
+    assert([1, 2, 3, 4].assumeSorted.quantile(0.75) == 3.5);
+
+    assert(sort([48, 31, 1, 37, 89, 83, 71]).quantile(0.5) == 48);
+}
+
+///
+@safe unittest
+{
+    import std.algorithm.sorting : sort;
+
+    assert([1, 2, 3, 4].sort!"a > b".quantile(0.25) == 1.5);
+    assert([1, 2, 3, 4].sort!"a > b".quantile(0.5) == 2.5);
+    assert([1, 2, 3, 4].sort!"a > b".quantile(0.75) == 3.5);
+
+    assert([1, 2, 3, 4].sort!"a^^2 < b^^2".quantile(0.25) == 1.5);
+    assert([1, 2, 3, 4].sort!"a^^2 < b^^2".quantile(0.5) == 2.5);
+    assert([1, 2, 3, 4].sort!"a^^2 < b^^2".quantile(0.75) == 3.5);
+}
+
+@safe unittest
+{
+    import std.internal.test.dummyrange;
+    import std.range : assumeSorted;
+    import std.math : approxEqual;
+    import std.stdio;
+
+    DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Forward) r;
+    assert(r.assumeSorted.quantile(0.25) == 3);
+    assert(r.assumeSorted.quantile(0.5).approxEqual(5.5));
+    assert(r.assumeSorted.quantile(0.75) == 8);
+}
+
+/**
+Convince function that wraps a call to the $(D quantile) function with 0.5 as
+the q parameter. For algorithmic complexity, refer to the docs of
+$(D quantile).
+
+Params:
+    r = a random access range or a forward range with a member length
+
+Returns:
+    the median of r
+*/
+auto median(alias less = "a < b", R)(SortedRange!(R, less) r)
+{
+    return quantile(r, 0.5);
+}
+
+///
+@safe unittest
+{
+    import std.algorithm.sorting : sort;
+    import std.range : assumeSorted;
+    import std.math : approxEqual;
+
+    assert(sort([48, 31, 1, 37, 89, 83, 71]).median == 48);
+
+    assert([3, 5, 12].assumeSorted.median == 5);
+    assert([3, 5, 7, 13, 21, 23, 23, 39].assumeSorted.median == 17);
+    assert([2.9, 6.2, 6.5, 20.5].assumeSorted.median.approxEqual(6.35));
+    assert([2.9, 5.8, 6.2, 6.5, 20.5].assumeSorted.median.approxEqual(6.2));
+
+    assert([12, 5, 3].assumeSorted!"a > b".median == 5);
+}
+
+version(release)
+{
+    @safe @nogc unittest
+    {
+        assert([1, 2][0 .. 0].assumeSorted.median == double.nan);
```

Comment:
I don't like this. A function should not have different behaviour based on whether it's compiled in release mode or not. If you think that `median` should return NaN for empty ranges, then it should return NaN for empty ranges, not do one thing in release mode, and something else in non-release mode. If you think it should assert, then it should assert (and leave release mode to UB, because a broken assertion means UB in release mode). Either that, or it should `assert(0)` with an appropriate error message.

Either way, the function should be consistent. It shouldn't assert in non-release mode yet return NaN in release mode.